src/lexer.coffee

```diff
@@ -1253,19 +1253,20 @@ REGEX_FLAGS  = /^\w*/
 VALID_FLAGS  = /^(?!.*(.).*\1)[imguy]*$/
 
 HEREGEX      = /// ^
-      (?:
-        [^\\/#]   |
-        \\[\s\S]  |
-        /(?!//)   |
-        # The comment should consume everything until the end of the line, including the '///'
-        \#(?!\{)(?:[^\n]*[\n])
-      )*
+  (?:
+      [^\\/#]   |
+      \\[\s\S]  |
+      /(?!//)   |
+      \\\#      | # Escaped hash.
+      # The comment should consume everything until the end of the line, including the '///'.
+      \#(?!\{)[^\n]*\n
+  )*
 ///
 
 HEREGEX_OMIT = ///
     ((?:\\\\)+)     # Consume (and preserve) an even number of backslashes.
   | \\(\s)          # Preserve escaped whitespace.
-  | \s+(?:#.*)?     # Remove whitespace and comments.
+  | \s+(?:\#.*)?     # Remove whitespace and comments.
```

Comment:
Ah, I had forgotten about this. Comments in heregex must be preceded by at least one whitespace. I think the correct fix is to revert this commit and adjust HEREGEX.

Reply:
Something like this:

```coffee
HEREGEX = /// ^
  (?:
    [^\\/#\s] |
    \\[\s\S]  |
    /(?!//)   |
    \#(?!\{)  |
    # Comments consume everything until the end of the line, including `///`.
    \s+(?:#(?!\{)[^\n]*[\n])? |
  )*
///
```

Reply:
So, hash preceded with the whitespace is a comment, e.g. no need to be escaped?

Reply:
I'm not sure about the whitespace. Is this valid heregex?
```coffeescript
/// ^(?: [^\\"\#] | \\[\s\S] |# (?!\{) ///
```

Reply:
Yes, I’d say that’s a valid regex.

`[$#]` – literal hash sign, because no space before.
`start#end` – same thing here.
`start #end` – this is a comment, because there’s a space before
`start \#end` – literal hash sign, because the hash is escaped.

This one should work and be treated as a comment since the hash is preceded by a newline (which is a whitespace character):

```coffee
///
[a-z]
# comment
\s+
```

However, my proposed regex currently doesn’t handle this case (hash sign preceded by nothing):

```coffee
///# comment
[a-z]///
```

Reply:
Something like this should handle the “comment at the very beginning” case:

```coffee
HEREGEX = /// ^
  (?:
    [^\\/#\s] |
    \\[\s\S]  |
    /(?!//)   |
    \#(?!\{)  |
    # Comments consume everything until the end of the line, including `///`.
    (?:\s+|^) (?: #(?!\{) [^\n]*[\n] )? |
  )*
///
```

Reply:
Also, if you look at the line I commented on you can see that I’m lot making this up. `REGEX_OMIT` only strips comments if preceded by a space.

Reply:
Hmm, since `REGEX_OMIT` requires a space before comments, I don’t think comments at the very beginning of a heregex has ever worked. So perhaps no need to worry about that. Or should we consider that a bug and fix that as well?

Reply:
It seems that order of the patterns matters. I guess it's some kind of "first come, first served" rule. The pattern for heregex comment has to be placed as the first option.
And the `[\n]` is also removed from the pattern.
```coffeescript
HEREGEX      = /// ^
  (?:
    # Comments consume everything until the end of the line, including `///`.
    (?:\s+|^)(?:#(?!\{)[^\n]*)? |
    [^\\/#]                     |
    \\[\s\S]                    |
    /(?!//)                     |
    \#(?!\{)
  )*
///
```

Reply:
I guess to cover the case when hash starts immediately after the `///`, we also have to adjust `REGEX_OMIT`.

Reply:
Correcting myself:
> The pattern for heregex comment has to be placed as the first option.

The position of heregex comment pattern can be anywhere as long as it is before this pattern `[^\\/#] ` which will match `\s` required by comment pattern.

> [\n] is also removed from the pattern.

Since we're matching `[^\n]*` (everything not newline) there is no need to specify `[\n]` explicitly at the end.

So, here is the improved (I hope) `HEREGEX`.
```coffeescript
HEREGEX      = /// ^
  (?:
      # Skip `\`, `/`, `#` and `\s` characters.
      [^\\/#\s]
      # Match `\` followed by any character.
    | \\[\s\S]
      # Match any `/` except `///`.
    | /(?!//)
      # Match `#` which is not part of interpolation, e.g. `#{}`.
    | \#(?!\{)
      # Comments consume everything until the end of the line, including `///`.
    | (?:\s+)(?:#(?!\{)[^\n]*)?
  )*
///
```